spec/ics-2-consensus-verification/README.md

```diff
@@ -0,0 +1,310 @@
+It generatess a unique list of headers starting from a genesis `ConsensusState` with arbitrary
+messages.
+
+`Blockchain` is defined as
+```go
+type Blockchain struct {
+  Verifier Verifier
+  Genesis ConsensusState
+  Commit Commit
+}
+```
+where
+  * `Verifier` is the lightclient verifier
+  * `Genesis` is the genesis `ConsensusState`
+  * `Commit` is the header generating function
+
+The headers generated from the `Blockchain` is expected to satisfy the followings:
+
+1. The `Header`s have no more than one direct child
+
+* Satisfied if: deterministic safety
+* Possible violation scenario: validator double signing, chain reorganization (Nakamoto consensus)
+
+2. The `Header`s eventually have at least one direct child
+
+* Satisfied if: liveness, light-client verifier continuity
+* Possible violation scenario: synchronised halt, incompatible hard fork
+
+3. The `Header`s are generated from the `Commit`, which ensures valid transition of the state
+
+* Satisfied if: correct block generation & state machine
+* Possible violation scenario: invariant break, validator cartel
+
+In these cases it is ensured that the `Verifier` works as intended.
+
+##### LightClient
+
+LightClient is defined as
+```go
+type LightClient struct {
+  Verifier Verifier
+  ConsensusState ConsensusState
+}
+```
+where
+  * `ConsensusState` is the root of trust providing the `VerifierBase`
+  * `Verifier` is the lightclient verification logic
+
+The exact type of each fields are depending on the type of the actual consensus logic.
+
+##### Functions
+
+* `Height :: Blockchain -> Header -> Uint` returns the position of the header in the
+  Blockchain's header list.
+
+#### Requirements
+
+Consensus verification requires the following accumulator primitives with datastructures and
+properties as defined in ICS23:
+
+* `AccumulatorRoot`
+
+#### Subprotocols
+
+##### Getter/Setter
+
+Identifiers are used to point the LightClients.
+
+```go
+function getLightClient(id :: Identifier) LightClient { /*impl specific*/ }
+
+function setLightClient(id :: Identifier, lc :: LightClient) { /*impl specific*/ }
+```
+
+##### Register
+The method that the identifiers are given to the lightclients(`newLightClientID`)
+is not defined by the protocol, rather depending on the implementation. Registering
+a new LightClient is expected to be permissionless.
+
+```go
+type RegisterLightClient LightClient
+
+function register(register :: RegisterLightClient) Identifier {
+  id := newLightClientID()
+  setLightClient(id, register)
+  return id
+}
+```
+
+##### Update
+
+Updating lightclient is done by submitting new header. The id is used to point the
+store lightclient that the logic will update. When the new header is verifiable with
+the stored lightclient's verifier and consensus state, then it
+
+```go
+type UpdateLightClient struct {
+  ID bytes
+  Header Header
+}
+
+function update(update :: UpdateLightClient) returns Maybe<Error> {
```

Comment:
Functions should be in Coffeescript. See ICS 1 - https://github.com/cosmos/ics/tree/master/spec/ics-1-ics-standard#pseudocode